### PR TITLE
Immutability rules

### PR DESCRIPTION
**Pull request by margarita-nedzelska-sonarsource**

### CODE REVIEW

**Comment by margarita-nedzelska-sonarsource** [comment]
After approval, I will take care of keeping nicely separated and green commits.

File: java-checks/src/main/java/org/sonar/java/checks/MutableMembersUsageCheck.java [hunk 1]
```diff
@@ -52,7 +52,7 @@ public class MutableMembersUsageCheck extends BaseTreeVisitor implements JavaFil
   private static final List<String> MUTABLE_TYPES = Arrays.asList(
     "java.util.Collection",
     "java.util.Date",
-    "java.util.Hashtable");
+    "java.util.Map");
```
Comment (dorian-burihabwa-sonarsource) [inline java-checks/src/main/java/org/sonar/java/checks/MutableMembersUsageCheck.java:55]: Becuase we now raise on `Map`, we should add `com.google.common.collect.ImmutableMap` to the IMMUTABLE_TYPES list.

**Review by dorian-burihabwa-sonarsource — CHANGES_REQUESTED** [review]
The PR looks good overall. I think we should revisit the tests to enforce the Java 9 constraints and consider ImmutableMap from guava to reduce the noise in the ruling.
(I tried to go through the whole ruling but it is way too noisy to check thoroughly)

File: java-checks/src/main/java/org/sonar/java/checks/MutableMembersUsageCheck.java (hunk 1 above)
Comment (margarita-nedzelska-sonarsource) [reply java-checks/src/main/java/org/sonar/java/checks/MutableMembersUsageCheck.java:55]: Yes, I was sure it's already included 🤔

File: java-checks/src/test/java/org/sonar/java/checks/PublicStaticMutableMembersCheckTest.java [hunk 2]
```diff
@@ -23,15 +23,22 @@
...
-      .onFile(testSourcesPath("checks/PublicStaticMutableMembersCheck.java"))
+      .onFile(mainCodeSourcesPath("checks/PublicStaticMutableMembersCheck.java"))
+      .withCheck(new PublicStaticMutableMembersCheck())
+      .verifyIssues();
+  }
+
+  @Test
+  void test_Java9() {
+    CheckVerifier.newVerifier()
+      .onFile(mainCodeSourcesPath("checks/PublicStaticMutableMembersCheckJava9.java"))
```
[the remark this answers is not in the document]
Comment (margarita-nedzelska-sonarsource) [reply java-checks/src/test/java/org/sonar/java/checks/PublicStaticMutableMembersCheckTest.java:41]: I don't think so, as the check itself doesn't care about the version.

These methods don't exist in Java less than 9, so won't be triggered anyway. I just moved them from a non-compiling folder to a compiling one, as we're on Java 17 now. Makes sense just to add its content to the existing file.

File: java-checks/src/test/java/org/sonar/java/checks/MutableMembersUsageCheckTest.java [hunk 3]
```diff
@@ -22,25 +22,32 @@
...
+      .onFile(mainCodeSourcesPath("checks/MutableMembersUsageCheck.java"))
       .withCheck(new MutableMembersUsageCheck())
       .verifyIssues();
   }
 
   @Test
   void test_java_9() {
     CheckVerifier.newVerifier()
-      .onFile(nonCompilingTestSourcesPath("checks/MutableMembersUsageCheck.java"))
+      .onFile(mainCodeSourcesPath("checks/MutableMembersUsageCheckJava9.java"))
```
[the remark this answers is not in the document]
Comment (margarita-nedzelska-sonarsource) [reply java-checks/src/test/java/org/sonar/java/checks/MutableMembersUsageCheckTest.java:40]: Initially, Java 9 methods were in a non-compiling file, as we used Java8. Now as we're using Java 17 to build we can put such test cases in the compiling folder. So I just moved the file. And left only non-compiling thongs in the non-compiling folder.

Do you suggest merging `MutableMembersUsageCheck.java` and `MutableMembersUsageCheckJava9.java`? I was thinking about it and maybe it's a better idea, but I'm not sure I understood what you mean by `sting again the same cases` Java 9 file targets methods `List.of`, `Set.of`, `List.copyOf`, `Set.copyOf` and there are no such methods in a `MutableMembersUsageCheck.java`

File: java-checks/src/main/java/org/sonar/java/checks/PublicStaticMutableMembersCheck.java [hunk 4]
```diff
@@ -83,9 +83,23 @@ public class PublicStaticMutableMembersCheck extends IssuableSubscriptionVisitor
         "org.apache.commons.collections4.list.UnmodifiableList")
       .names(DECORATE)
       .withAnyParameters()
-      .build()
+      .build(),
+    MethodMatchers.create().ofTypes("com.google.common.collect.Sets").names("union", "intersection", "difference", "symmetricDifference").withAnyParameters().build(),
```
[the remark this answers is not in the document]
Comment (margarita-nedzelska-sonarsource) [reply java-checks/src/main/java/org/sonar/java/checks/PublicStaticMutableMembersCheck.java:87]: I think all these methods are covered by these 2 matchers ` MethodMatchers.create().ofType(type -> MutableMembersUsageCheck.containsImmutableLikeTerm(type.name())).anyName().withAnyParameters().build(),
    MethodMatchers.create().ofAnyType().name(MutableMembersUsageCheck::containsImmutableLikeTerm).withAnyParameters().build(),`

File: java-checks-test-sources/src/main/java/checks/MutableMembersUsageCheck.java [hunk 5]
```diff
@@ -26,6 +33,93 @@ class MutableMembersUsageCheck {
...
+    return immutableMap; // Compliant
+  }
+
+  private Map<String, String> immutableMap2 = ImmutableMap.of("abc", "abc");
+
+  public Map<String, String> getImmutableMap2() {
+    return immutableMap2; // Compliant
+  }
+
+  private static final Set<String> unmodifiableFromStream = // FP
```
Comment (dorian-burihabwa-sonarsource) [inline java-checks-test-sources/src/main/java/checks/MutableMembersUsageCheck.java:54]: Are this and the following field still FPs?

File: java-checks/src/main/java/org/sonar/java/checks/PublicStaticMutableMembersCheck.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (dorian-burihabwa-sonarsource) [reply java-checks/src/main/java/org/sonar/java/checks/PublicStaticMutableMembersCheck.java:87]: Indeed. Thanks for the clarification

**Review by dorian-burihabwa-sonarsource — APPROVED** [review]
Thanks for all the changes. I am not sure about the leftover FP comments in the test sources but the changes overall look good to me. I will let you handle the merge as you see fit 👍🏿

File: java-checks/src/test/java/org/sonar/java/checks/PublicStaticMutableMembersCheckTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (dorian-burihabwa-sonarsource) [reply java-checks/src/test/java/org/sonar/java/checks/PublicStaticMutableMembersCheckTest.java]: Thanks for clarifying

File: java-checks-test-sources/src/main/java/checks/MutableMembersUsageCheck.java (hunk 5 above)
Comment (margarita-nedzelska-sonarsource) [reply java-checks-test-sources/src/main/java/checks/MutableMembersUsageCheck.java:54]: Oh, probably leftovers. Will remove

**Pull request merged**